Code Review: Smart Contract VM 02/13/19

kantai opened the pull request:
This PR adds the following:

1. Contract interface for executing a transaction within the context of a transaction, with an in-memory implementation of that interface.
1. `(define-public ...)` function for specifying the public functions in a contract.
1. Support for buffer literals via hex strings or ascii string literals.
1. Type definitions that use the structural representation described in SIP-002 (i.e., `(list 3 (buffer 5))`) 
1. Max value size enforcement
1. Max stack and context depth enforcement

This PR also reimplements the lexer from #908 so that it is a much more standard lexer (and the lexer itself handles literals).

This PR addresses:

#911 
#913 -- though this will need to be revisited during the course of the implementation
#914 
#915 
#917 
#918

kantai commented:
Pushed some more work into this PR:

Adding type parameters to (define-public ...): #922 

- Use loose type admission for maximum buffer/tuple size: basically, we want it so that we enforce a maximum buffer size, but will accept smaller buffers (the containing type is just the maximum of its composite types)

Adding initial implementation of a "Principal" type, parsed as described in SIP-002. I got a little carried away on this and implemented C32 address decoding -- see the implementation in `src/address/c32.rs`

jcnelson commented on src/vm/callables.rs:
```diff
@@ -0,0 +1,84 @@
+                }
+            }
+        } else {
+            let types = self.types.as_ref().unwrap(); // if types is None, and is_public = true, we should panic.
```
Totally fine that this panics, but could you insert an error message here as well?

jcnelson replied:
Alternatively, you could make this type of function unrepresentable by removing `is_public` from `DefinedFunction` and instead create a type enumeration like this:

```
pub enum FunctionTypes {
   PublicFunction(DefinedFunction),
   PrivateFunction(DefinedFunction)
}
```

Then, you'd simply handle this as a `match` statement.

kantai replied:
Yep, I'll create a new type for this

jcnelson commented on src/vm/contexts.rs:
```diff
@@ -21,26 +28,38 @@ impl <'a> Environment <'a> {
 
+// Aaron: note -- only the global context will ever have DefinedFunctions
+//        so it is probably worthwhile to separate into 2 types.
+
```
Probably a good idea -- goes along with the general Rust-ism of avoiding non-sensical instantiations by making them unrepresentable in the first place.

kantai replied:
Yep

jcnelson commented on src/vm/errors.rs:
```diff
@@ -13,24 +13,48 @@ pub enum Error {
+    BadSymbolicRepresentation(String),
+    ReservedName(String),
+    InterpreterError(String),
+    MultiplyDefined(String)
```
"MultiplyDefined"?  As in, defined multiple times, or something to do with the multiplication operator?

kantai replied:
Ah, it's an error when you use the same variable name multiple times in a variable declaration like, the function definition:

`(define (foo a a) (+ a a))`

I can rename it to VariableDefinedMultipleTimes

jcnelson commented on src/vm/types.rs:
```diff
@@ -1,66 +1,208 @@
-    Buffer(Box<[char]>),
+    Buffer(BuffData),
     List(Vec<Value>, TypeSignature),
+    Principal(u8, Vec<u8>), // a principal is a version byte + hash160 (20 bytes)
```
You can enforce this in the type system by using `[u8; 20]` instead of `Vec<u8>`

jcnelson commented on src/vm/types.rs:
```diff
@@ -70,143 +212,399 @@ impl TupleTypeSignature {
+        let mut value_size: i128 = 0;
+        for (name, type_signature) in self.type_map.iter() {
+            // we only accept ascii names, so 1 char = 1 byte.
+            name_size = name_size.checked_add(name.len() as i128).unwrap();
```
For these `unwrap()`s, can you add an error message?

jcnelson replied:
Also, how certain are we that these `unwrap()`s never occur?

kantai replied:
Sure -- will add an error message.

These panics should never occur, since the composite values of the tuple should be smaller than MAX_VALUE_SIZE (much smaller than i128::max bytes) and combining them until the tuple is too large would require an extraordinarily large program. However, since the max value size check for tuple (and list) types occurs using the `size()` method, it's theoretically possible. So I'll change these to raise a ValueTooLarge error.

jcnelson commented on src/vm/types.rs:
```diff
@@ -70,143 +212,399 @@ impl TupleTypeSignature {
+        value_size.checked_add(name_total_size).unwrap()
+    }
+
+    // NOTE: this function mutates self _even if it returns an error_.
```
Is doing so necessary?  Is there a risk that this object is left in a partially-computed state?  It might be safer just to clone `self`, apply the `expand_to_admit` method on the copy, and then if everything works, apply the changes to `self`.

kantai replied:
The expand_to_admit function is only ever called to build a new type, so there's no risk at the moment of using an object that was in the partially computed state.

jcnelson replied:
Gotcha.  We'll want to make sure it stays that way.

jcnelson commented on src/vm/types.rs:
```diff
@@ -70,143 +212,399 @@ impl TupleTypeSignature {
+        }
+    }
+
+    pub fn get_empty_list_type() -> TypeSignature {
```
Why is the `atomic_type` for an empty list an `IntType`?

kantai replied:
an empty list in this form has a maximum length of 0, meaning that it can't ever hold any values whatsoever. It should really be something like an "Any" type, though, so that if you had, say, a map value that required something of type `(list 1 bool)`, an empty list would fit, and it should fit in `(list 1 int)` as well.

jcnelson replied:
Okay, sounds good.

jcnelson commented on src/vm/types.rs:
```diff
@@ -70,143 +212,399 @@ impl TupleTypeSignature {
+    //     e.g.: (list "abcd" "abc") will currently error because one etry is
+    //           if type (buffer 4) and the other is of type (buffer 3)
+    //       my feeling is that this should probably be allowed, and the resulting
+    //       type should be (list 2 (buffer 4)) 
```
I agree, just as long as we can make it clear somehow that `(list "abcdefghijk" "abc")` costs as much to process as `(list "abcdefghijk" "        abc")`.

kantai replied:
yep -- we'd need to make this clear. I think the behavior we want is that for any functions which accepts two buffers, to treat the smaller buffer as left padded with zeros. currently we don't have _any_ buffer functions (except `eq?`).

jcnelson replied:
How do you want to handle `eq?` for buffers that get implicitly left-padded with zeros?  I don't think `\x00abc` and `abc` are necessarily equal.  Maybe the buffer type should include a `length` field internally to avoid this problem?

kantai replied:
Yeah, I was just going to comment on this. Those shouldn't be equal. Buffers should have a length, which they do currently (via their `vec.len()`), but the Buffer type has a _maximum length_, which is a little silly for a buffer's runtime type (because its length is always equal to its maximum length), but for declared types like public function inputs and datamap keys, values, it makes sense, because the maximum length determines whether or not a given runtime value would be admissable.

kantai replied:
(this contradicts my previous response up there --- I don't think we should be automatically extending buffers. if we decide to include buffer functions like `strncat`, they should specify their padding behavior, if they have any)

jcnelson commented on src/vm/parser.rs:
```diff
@@ -1,103 +1,165 @@
+    //    lazy_static (or just hand implementing that), and I'm not convinced
+    //    it's worth either (1) an extern macro, or (2) the complexity of hand implementing.
+    let lex_matchers: &[LexMatcher] = &[
+        LexMatcher::new(r##""(?P<value>((\\")|([[:ascii:]&&[^"\n\r\t]]))*)""##, TokenType::StringLiteral),
```
What is `:ascii:` here?  Is it _printable_ ASCII characters?  Also, can we have a lexer test below for ensuring that a code snippit with non-printable characters does not parse?

jcnelson replied:
Also, the `str` type is a _unicode_ string slice.  We'll want the lexer to reject non-ASCII strings.

kantai replied:
Yeah -- that `:ascii:` class should actually be `:print:` for printable ascii characters.

But this lexer _will_ reject non-ASCII strings. None of these matchers will match a non-ascii character, which will cause the lexer to fail (if the lexer cannot process a given character, it exhaust all the matches and returns a ParseError)

kantai commented:
Okay -- I pushed changes that addressed all of those issues.

1. Use two types for Global vs. Local contexts.
2. Use two types for public vs. private defined functions.
3. Raise a ValueTooLarge error in the event of an i128 overflow in type size.
4. Use printable ascii character class instead of ascii character class (and add unicode rejection test)
5. Use [u8;20] for the principal data
6. Rename MultiplyDefined error
7. Implement value equality check which checks for _data equality_ not type equality. Will create a new issue for speccing out buffer functions.

jcnelson commented:
Thanks!  As soon as CircleCI finishes, please go ahead and merge to `develop`.  Thanks also for implementing `c32.rs` :grinning:

kantai commented:
Thanks for the review! The Circle CI tests look like they've finished successfully (https://circleci.com/gh/blockstack/blockstack-core/1716), so I'll go ahead and merge (Circle github notifications/commit check updates appear to currently be broken).

blockstack-devops commented:
This pull request has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
